packages/react-dom/src/__tests__/ReactDOMTestSelectors-test.internal.js

```diff
@@ -0,0 +1,1427 @@
+      function Outer() {
+        return (
+          <div id="outer">
+            <Inner />
+          </div>
+        );
+      }
+
+      function Inner() {
+        return <div id="inner" />;
+      }
+
+      render(<Outer />, container);
+
+      const matches = findAllNodes(container, [createComponentSelector(Outer)]);
+      expect(matches).toHaveLength(1);
+      expect(matches[0].id).toBe('outer');
+
+      expect(() => {
+        findAllNodes(matches[0], [
+          createComponentSelector(Inner),
+          createTestNameSelector('inner'),
+        ]);
+      }).toThrow(
+        'Invalid host root specified. Should be either a React container or a node with a testname attribute.',
+      );
+    });
+
+    it('should support an multiple component types in the selector array', () => {
+      function Outer() {
+        return (
+          <>
+            <div data-testname="match" id="match1" />
+            <Middle />
+          </>
+        );
+      }
+      function Middle() {
+        return (
+          <>
+            <div data-testname="match" id="match2" />
+            <Inner />
+          </>
+        );
+      }
+      function Inner() {
+        return (
+          <>
+            <div data-testname="match" id="match3" />
+          </>
+        );
+      }
+
+      render(<Outer />, container);
+
+      let matches = findAllNodes(document.body, [
+        createComponentSelector(Outer),
+        createComponentSelector(Middle),
+        createTestNameSelector('match'),
+      ]);
+      expect(matches).toHaveLength(2);
+      expect(matches.map(m => m.id).sort()).toEqual(['match2', 'match3']);
+
+      matches = findAllNodes(document.body, [
+        createComponentSelector(Outer),
+        createComponentSelector(Middle),
+        createComponentSelector(Inner),
+        createTestNameSelector('match'),
+      ]);
+      expect(matches).toHaveLength(1);
+      expect(matches[0].id).toBe('match3');
+
+      matches = findAllNodes(document.body, [
+        createComponentSelector(Outer),
+        createComponentSelector(Inner),
+        createTestNameSelector('match'),
+      ]);
+      expect(matches).toHaveLength(1);
+      expect(matches[0].id).toBe('match3');
+    });
+
+    it('should find multiple matches', () => {
+      function Example1() {
+        return (
+          <div>
+            <div data-testname="match" id="match1" />
+          </div>
+        );
+      }
+
+      function Example2() {
+        return (
+          <div>
+            <div data-testname="match" id="match2" />
+            <div data-testname="match" id="match3" />
+          </div>
+        );
+      }
+
+      render(
+        <>
+          <Example1 />
+          <Example2 />
+        </>,
+        container,
+      );
+
+      const matches = findAllNodes(document.body, [
+        createTestNameSelector('match'),
+      ]);
+      expect(matches).toHaveLength(3);
+      expect(matches.map(m => m.id).sort()).toEqual([
```

Comment:
This is not necessarily a concern with the implementation but if this lands the documentation should warn that we shouldn't rely on the order since it's an implementation detail of findAllNodes. I guess technically this is the order in the react tree which isn't the same as in the DOM tree because e.g. portals?